File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str



@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	# TODO: implement
	print(recipeName)
	temp:str = re.sub(r'[-_]', ' ', recipeName)
	print(temp)
	temp = re.sub(r'[^a-zA-Z ]', '', temp)
	print(temp)
	temp = re.sub(r'\s+', ' ', temp).strip().title()
	print(temp)
	recipeName = temp
	
	print(cookbook)

	# Check if the recipeName is valid
	if len(recipeName) == 0:
		return None
	
	return recipeName


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	# TODO: implement me
	data = request.get_json()
	print(data)
	food_type = data.get('type', '')
	food_name = data.get('name', '')
	required_items = data.get('requiredItems', [])

	if food_type not in ['recipe', 'ingredient']:
		return 'Invalid type: must be "recipe" or "ingredient"', 400
	
	cook_time = -1
	if food_type == 'ingredient':
		cook_time = data.get('cookTime', -1)
		if not isinstance(cook_time, int) or cook_time < 0:
			return 'Invalid cookTime: must be a greater than or equal to 0', 400

	
	if food_name in cookbook:
		return 'Invalid name: must be unique', 400
	
	seen_items = set()
	list_items = []
	for item in required_items:
		if item['name'] in seen_items:
			# Note: Original Auotests does not check for duplicate item names
			return 'Invalid requiredItems: duplicate item names are not allowed', 400
		seen_items.add(item['name'])
		list_items.append(RequiredItem(name=item['name'], quantity=item['quantity']))

	if food_type == 'recipe':
		cookbook[food_name] = Recipe(name=food_name, required_items=list_items)
	elif food_type == 'ingredient':
		cookbook[food_name] = Ingredient(name=food_name, cook_time=cook_time)

	# Autotests needs empty object to be returned when succesfully created
	# return jsonify({'msg': f'Created {food_type} "{food_name}"'}), 200
	return jsonify({}), 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	# TODO: implement me
	return 'not implemented', 500


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)

File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Map<string, cookbookEntry> = new Map(); // cookbookEntry = null;

// Task 1 helper (don't touch)
app.post("/parse", (req: Request, res: Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  }
  res.json({ msg: parsed_string });
  return;

});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {

  recipeName = recipeName.replace(/[-_]/g, " ");

  recipeName = recipeName.replace(/[^a-zA-Z ]/g, "");

  recipeName = recipeName.replace(/\s+/g, " ");

  recipeName = recipeName.toLowerCase();

  console.log(recipeName);
  // Capitalize the first letter of each word
  recipeName = recipeName.split(" ").map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(" ");

  console.log(recipeName);

  if (recipeName.length == 0) {
    return null;
  }

  const mapAsObject = Object.fromEntries(cookbook);
  console.log(JSON.stringify(mapAsObject, null, 2));

  return recipeName
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req: Request, res: Response) => {

  const data = req.body;
  console.log(data);
  const food_name: string = data.name;
  const food_type: string = data.type;

  if (food_type != "ingredient" && food_type != "recipe") {
    res.status(400).send("invalid type");
    return;
  }

  let cook_time = -1;
  if (food_type == "ingredient") {
    cook_time = data.cookTime;
    if (cook_time < 0) {
      res.status(400).send("invalid cook time");
      return;
    }
  }

  if (cookbook.has(food_name)) {
    res.status(400).send("entry already exists");
    return;
  }

  let list_items: requiredItem[] = [];
  let seen_array: string[] = [];
  if (food_type == "recipe") {
    for (let item of data.requiredItems) {

      if (item.name in seen_array) {
        res.status(400).send("duplicate item");
        return;
      }

      if (item.quantity < 0) {
        res.status(400).send("invalid quantity");
        return;
      }

      let required_item: requiredItem = {
        name: item.name,
        quantity: item.quantity
      }

      seen_array.push(item.name);
      list_items.push(required_item);
    }
  }

  if (food_type == "ingredient") {
    const new_ingredient: ingredient = {
      name: food_name,
      type: food_type,
      cookTime: cook_time
    }
    cookbook.set(food_name, new_ingredient);
  } else {
    const new_recipe: recipe = {
      name: food_name,
      type: food_type,
      requiredItems: list_items
    }
    cookbook.set(food_name, new_recipe);
  }
  res.status(200).send({});
});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req: Request, res: Request) => {
  const foodSummaryName = (req.query.name as string);
  const ingredientSummary: Map<string, number> = new Map();
  let totalCookTime = 0;

  // Using a standard array as a queue
  const foodQueue: string[] = [foodSummaryName];

  while (foodQueue.length > 0) {
      const currentFoodName:string = foodQueue.shift()!; // popleft equivalent

      if (!cookbook.has(currentFoodName)) {
          return res.status(400).send(`Invalid recipe name: "${currentFoodName}"`);
      }

      const entry = cookbook.get(currentFoodName)! as recipe;

      if (entry.type === "ingredient") {
          return res.status(400).send(`Invalid recipe name: "${currentFoodName}" is an ingredient, not a recipe`);
      }

      // Assuming Quantity is always non-negative integer
      for (const item of entry.requiredItems) {
        let isRecipe = cookbook.get(item.name)?.type === "recipe";
          if (isRecipe) {
              for (let i = 0; i < item.quantity; i++) {
                  foodQueue.push(item.name);
              }
          } else {
              if (!cookbook.has(item.name)) {
                  return res.status(400).send(`Invalid ingredient name: "${item.name}"`);
              }

              const currentIngredient = cookbook.get(item.name)! as ingredient;

              ingredientSummary.set(item.name, (ingredientSummary.get(item.name) || 0) + item.quantity);
              totalCookTime += currentIngredient.cookTime * item.quantity;
          }
      }
  }

  const result = {
      name: foodSummaryName,
      cookTime: totalCookTime,
      ingredients: Object.fromEntries(ingredientSummary)
  };

  console.log(result);
  return res.status(200).json(result);

});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});
